### ISSUE TITLE
Pproperty ApiKeySourceType not defined for resource of type AWS::Serverless::Api

### ISSUE THREAD

**Issue by imjustd** [awslabs/serverless-application-model#1274]
**Description:**
I wanted to specify  the Api Key Source to `AUTHORIZER` in the SAM `template.yaml` file since I use custom Lambda authorizer but it didn't succeed.

**Steps to reproduce the issue:**
1. Create new SAM project and try to add a new resource that uses `AWS::Serverless::Api` similar to:
```
Resources:
  MyApi:
    Type: AWS::Serverless::Api
    Properties:
      ApiKeySourceType: AUTHORIZER
      StageName: Prod
```
2. run `sam validate`

**Observed result:**
When running `sam validate` I got the error:
```
Error: [InvalidResourceException('MyApi', 'property ApiKeySourceType not defined for resource of type AWS::Serverless::Api')] ('MyApi', 'property ApiKeySourceType not defined for resource of type AWS::Serverless::Api')
```

**Expected result:**
- I expect `AWS::Serverless::Api` to accept this parameter `ApiKeySourceType` because it's supported in the `AWS::ApiGateway::RestApi` https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-apigateway-restapi.html#cfn-apigateway-restapi-apikeysourcetype

**Note:** I tried to use `AWS::ApiGateway::RestApi` to define my API in the SAM template but it didn't work out.

**Comment by imjustd** [comment]
Any update on this?

**Comment by pedrofurlanetto90** [comment]
Is there any update on this? Without this parameter, I needed to use swagger definitions on API Gateway

**Comment by imjustd** [comment]
Almost a year has passed, do you have some update?

**Comment by pedrofurlanetto90** [comment]
We really need this feature. Do you have any update? Any information on why AWS::Serverless::Api does not support it?

@imjustd did you find some workaround?

**Comment by imjustd** [comment]
> @imjustd did you find some workaround?

Currenty I use this commands after deploy:
```
aws apigateway update-rest-api --rest-api-id <api_id> --patch-operations op=replace,path=/apiKeySource,value=AUTHORIZER && \
aws apigateway create-deployment --rest-api-id <api_id> --stage-name Prod
```
I would actually try to implement this myself if I know what needs to be changed in the SAM code.

**Comment by nayaranunes** [comment]
The property 'ApiKeySourceType' isn't defined for resource of type AWS::Serverless::Api, try this:
```Resources:
   MyApi:
      Type: AWS::Serverless::Api
      Properties:
         StageName: Prod
         Auth:
           DefaultAuthorizer: MyAuthorizer
           Authorizers:
             MyAuthorizer:
```
The ApiAuth doc can be helpful:
 https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-property-api-apiauth.html

**Comment by imjustd** [comment]
In the ApiAuth documentation I can't find a way to specify the key source type to `AUTHORIZER`. Still have to use the workaround posted above.

**Comment by nayaranunes** [comment]
It's seems that the property `ApiKeySourceType` is defined in type `AWS::Serverless::RestApi` not `AWS::Serverless::Api` 

![image](https://user-images.githubusercontent.com/39678494/108278739-354a7700-715a-11eb-9163-41a55c541462.png)

**Comment by imjustd** [comment]
Yes, I tried to use `RestApi` without success, so will stay with `AWS::Serverless::Api`

**Comment by nayaranunes** [comment]
In my test with RestApi the template was `valid` (screenshot above) and I don't got the error `InvalidResourceException('MyApi', 'property ApiKeySourceType not defined...`

**Comment by imjustd** [comment]
My question is: will the SAM lambda work same with `RestApi` as with `Api`? I will try to test it in the next days but was wondering if someone already tried this.

**Comment by nverino** [comment]
You can add an `x-amazon-apigateway-api-key-source` property with `AUTHORIZER` as value in your `DefinitionBody`

![image](https://user-images.githubusercontent.com/583720/109202660-fcc22300-7781-11eb-83e9-984e3592649c.png)

More details [here](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-swagger-extensions-api-key-source.html)

**Comment by imjustd** [comment]
> It's seems that the property `ApiKeySourceType` is defined in type `AWS::Serverless::RestApi` instead of `AWS::Serverless::Api`
> 
> ![image](https://user-images.githubusercontent.com/39678494/108278739-354a7700-715a-11eb-9163-41a55c541462.png)

I just tested this but it seems although the template is reported as valid `AWS::Serverless::RestApi` is not an existing resource for SAM template, so I get this error when trying to deploy it:
```
Template format error: Unrecognized resource types: [AWS::Serverless::RestApi]
```
Also can confirm this from the official AWS documentation (only Api and HttpApi are valid both not supporting the `ApiKeySourceType` property):
![image](https://user-images.githubusercontent.com/590260/117651759-9a7a9b00-b192-11eb-863f-ebbf512737ac.png)

**Comment by imjustd** [comment]
> You can use the OpenAPI extension `x-amazon-apigateway-api-key-source` with `AUTHORIZER` as value in your `DefinitionBody`
> 
> ![image](https://user-images.githubusercontent.com/583720/109202660-fcc22300-7781-11eb-83e9-984e3592649c.png)
> 
> More details [here](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-swagger-extensions-api-key-source.html)

Sorry, I can't make it work, but I'm against of using Swagger or workarounds for such a small thing.

I still think `ApiKeySourceType` needs to be implemented as part of the `AWS::Serverless::Api` resource, no workarounds, no using other resources as `ResApi` and `HttpApi`. If I can change this from AWS console why the template wouldn't allow it?

Still using my workaround:
```
aws apigateway update-rest-api --rest-api-id <api_id> --patch-operations op=replace,path=/apiKeySource,value=AUTHORIZER && \
aws apigateway create-deployment --rest-api-id <api_id> --stage-name Prod
```

**Comment by jfuss** [comment]
Updating labels. 

To the community: We prioritize issues base on Github reactions. I will flag this for pm-review, but reactions will help us gauge general community need. 

This does seem like something we can support and if someone is interested in contributing it, we would accept the PR.

**Comment by imjustd** [comment]
> To the community: We prioritize issues base on Github reactions. I will flag this for pm-review, but reactions will help us gauge general community need.

To the community: Please vote for this with thumbs-up reaction.

> This does seem like something we can support and if someone is interested in contributing it, we would accept the PR.

OK, we're clear.

**Comment by heitorlessa** [comment]
Just hit this issue with a customer, it'd be great to have supported.

**Customer use case**: Mobile App can't be updated for certain customers thus requiring API Keys to be dynamically assigned at the Authorizer level. This allows them to set rate limit and burst depending on their tenant group and without having the Mobile devices updated.

cc @praneetap

**Comment by NickHeap2** [comment]
Just adding my findings here:

If you change from AWS::Serverless::Api to AWS::Serverless::RestApi with ApiKeySourceType included then sam validate will tell you the template is valid but if you try to sam deploy it you will get a "Template format error: Unrecognized resource types: [AWS::Serverless::RestApi]" error.

I changing the code locally to allow ApiKeySourceType to be added to AWS::Serverless::Api which would then get copied by sam to AWS::Serverless::RestApi and that then passed sam validate but when trying to sam deploy it gave the error "property ApiKeySourceType not defined for resource of type AWS::Serverless::Api".

What I'd ideally like to do is be able to use an OpenAPI stub that can include x-amazon-apigateway-api-key-source and no paths and have sam add the paths for me. This looks to be currently blocked by the code here: [samtranslator/plugins/api/implicit_api_plugin.py#L208](https://github.com/aws/serverless-application-model/blob/ab6943a340a3f489af62b8c70c1366242b2887fe/samtranslator/plugins/api/implicit_api_plugin.py#L208)

A reasonable change may be that if there are no paths at all in the OpenAPI spec then to add in __MANAGE_SWAGGER so that paths will then be added in by sam.

**Comment by lee-11** [comment]
I ran into this yesterday and found this issue today.  The behavior I'm seeing (implied above, but I thought it would be helpful to say specifically) is that the `ApiKeySourceType` property is reset to `HEADER` after every deployment, so it's a real pain.  

Between the time the stack deployment updates the API and the call to the `cli` to update the `ApiKeySourceType` setting all requests to the API will get authorization errors since the `x-api-key` header isn't present (not needed when the Authorizer is doing the work of assigning the API key).  

So, in short, "bump" or "+1".

**Comment by moelasmar** [comment]
Thanks @lee-11 for your contribution. PR https://github.com/aws/serverless-application-model/pull/2298 got merged, and waiting for the next release.

**Comment by jfuss** [comment]
This was release in [v1.46](https://github.com/aws/serverless-application-model/releases/tag/v1.46.0).

closing
